controllers/storage.go

```diff
@@ -145,6 +146,123 @@ func (s *Storage) RemoveAllButCurrent(artifact sourcev1.Artifact) ([]string, err
 	return deletedFiles, nil
 }
 
+// getGarbageFiles returns all files that need to be garbage collected for the given artifact.
+// Garbage files are determined based on the below flow:
+// 1. collect all files with an expired ttl
+// 2. if we satisfy maxItemsToBeRetained, then return
+// 3. else, remove all files till the latest n files remain, where n=maxItemsToBeRetained
+func (s *Storage) getGarbageFiles(artifact sourcev1.Artifact, maxItemsToBeRetained int, ttl time.Duration) ([]string, error) {
+	localPath := s.LocalPath(artifact)
+	dir := filepath.Dir(localPath)
+	garbageFiles := []string{}
+	filesWithCreatedTs := make(map[time.Time]string)
+	// sortedPaths contain all files sorted according to their created ts.
+	sortedPaths := []string{}
+	now := time.Now().UTC()
+	totalFiles := 0
+	var errors []string
+	_ = filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
```

Comment:
I don't think we can do that here. The whole point of these garbage retention options, is that we make sure that there are a _specific_ number of files after each reconciler run. Adding a max amount of items can mislead users. If we want to avoid "process time starvation", we could cap `maxItemsToBeRetained` to a reasonable number, so that users don't abuse it.